server/src/main/java/io/pravega/schemaregistry/service/SchemaRegistryService.java

```diff
@@ -0,0 +1,924 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.schemaregistry.service;
+
+import com.fasterxml.jackson.databind.MapperFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.SerializationFeature;
+import com.fasterxml.jackson.module.jsonSchema.JsonSchema;
+import com.google.common.base.Charsets;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.protobuf.DescriptorProtos;
+import io.pravega.common.Exceptions;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.util.Retry;
+import io.pravega.schemaregistry.MapWithToken;
+import io.pravega.schemaregistry.common.FuturesCollector;
+import io.pravega.schemaregistry.common.NameUtil;
+import io.pravega.schemaregistry.contract.data.BackwardAndForward;
+import io.pravega.schemaregistry.contract.data.Compatibility;
+import io.pravega.schemaregistry.contract.data.EncodingId;
+import io.pravega.schemaregistry.contract.data.EncodingInfo;
+import io.pravega.schemaregistry.contract.data.GroupHistoryRecord;
+import io.pravega.schemaregistry.contract.data.GroupProperties;
+import io.pravega.schemaregistry.contract.data.SchemaInfo;
+import io.pravega.schemaregistry.contract.data.SchemaWithVersion;
+import io.pravega.schemaregistry.contract.data.SerializationFormat;
+import io.pravega.schemaregistry.contract.data.VersionInfo;
+import io.pravega.schemaregistry.exceptions.IncompatibleSchemaException;
+import io.pravega.schemaregistry.exceptions.PreconditionFailedException;
+import io.pravega.schemaregistry.exceptions.SerializationFormatMismatchException;
+import io.pravega.schemaregistry.rules.CompatibilityChecker;
+import io.pravega.schemaregistry.rules.CompatibilityCheckerFactory;
+import io.pravega.schemaregistry.storage.ContinuationToken;
+import io.pravega.schemaregistry.storage.SchemaStore;
+import io.pravega.schemaregistry.storage.StoreExceptions;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.avro.Schema;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+import java.util.AbstractMap;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+import static io.pravega.schemaregistry.contract.data.BackwardAndForward.*;
+
+/**
+ * Schema registry service backend.
+ */
+@Slf4j
+public class SchemaRegistryService {
+    private static final Retry.RetryAndThrowConditionally RETRY = Retry.withExpBackoff(1, 2, Integer.MAX_VALUE, 100)
+                                                                       .retryWhen(x -> Exceptions.unwrap(x) instanceof StoreExceptions.WriteConflictException);
```

Comment:
Is `WriteConflictException` a recoverable/transient failure? If not retrying wouldn't help. The comment `Exception type when you are attempting to update a stale value.` in the exception doesn't seem to suggest that it is a transient fault.

Also, curious why you used exponential backoff as the strategy for determining the retry interval. Its usage below doesn't seem to suggest that the operation is a background operation or the fault induced by throttling. A simple regular interval strategy might just work fine, and might even be better arguably.

Reply:
It is concurrency error as we rely on optimistic concurrency.. 
so doing a get and update again should work. 

Perhaps a random delay is better than fixed or exponential backoff delay.